lib/elixir/lib/calendar/datetime.ex

```diff
@@ -245,6 +440,149 @@ defmodule DateTime do
     end
   end
 
+  # Takes a datetime and in case it is is on the 61st second (:60) it will
+  # check if it is a known leap second.
+  # All datetimes with non ISO calendars return :ok
+  @spec validate_positive_leap_second(
+          Calendar.datetime(),
+          TimeZoneDatabaseClient.tz_db_or_config()
+        ) :: :ok | {:error, atom}
+  defp validate_positive_leap_second(%{second: second, calendar: calendar}, _)
+       when second != 60 or calendar != Calendar.ISO do
+    :ok
+  end
+
+  defp validate_positive_leap_second(dt, tz_db_or_config) do
+    utc_dt = to_zero_total_offset(dt)
+
+    case TimeZoneDatabaseClient.is_leap_second(utc_dt, tz_db_or_config) do
+      {:ok, true} ->
+        :ok
+
+      {:ok, false} ->
+        {:error, :invalid_leap_second}
+
+      {:error, _} = error ->
+        error
+    end
+  end
+
+  @doc """
+  Takes a `DateTime` and a time zone.
+
+  Returns a `DateTime` for the same point in time, but instead at the time zone
+  provided.
+
+  Requires passing a `TimeZoneDatabase` as an argument or setting it with
+  `TimeZoneDatabaseClient.set_database/1`.
+
+  ## Examples
+
+      iex> cph_datetime = DateTime.from_naive!(~N[2018-07-16 12:00:00], "Europe/Copenhagen", FakeTimeZoneDatabase)
+      iex> {:ok, pacific_datetime} = DateTime.shift_zone(cph_datetime, "America/Los_Angeles", FakeTimeZoneDatabase)
+      iex> pacific_datetime
+      #DateTime<2018-07-16 03:00:00-07:00 PDT America/Los_Angeles>
+
+  """
+  @doc since: "1.8.0-dev"
+  @spec shift_zone(t, Calendar.time_zone(), TimeZoneDatabaseClient.tz_db_or_config()) ::
+          {:ok, t} | {:error, :time_zone_not_found} | {:error, :incompatible_calendars}
+  def shift_zone(datetime, time_zone, tz_db_or_config \\ :from_config)
+
+  def shift_zone(%{time_zone: time_zone} = datetime, time_zone, _) do
+    # When the desired time_zone is the same as the existing time_zone just
+    # return the passed datetime unchanged.
+    {:ok, datetime}
+  end
+
+  def shift_zone(
+        %{calendar: Calendar.ISO} = datetime,
+        time_zone,
+        tz_db_or_config
+      ) do
+    in_utc = to_zero_total_offset(datetime)
+
+    case TimeZoneDatabaseClient.by_utc(in_utc, time_zone, tz_db_or_config) do
+      {:ok, period} ->
+        naive_datetime =
+          in_utc
+          |> naive_add_preserve_second_60(period.utc_offset + period.std_offset)
```

Comment:
I think this is very inefficient. to_zero_total_offset converts to iso days, then back to naive, then we get the period, convert back to iso days, adds the offsets, and convert back to a datetime again.